### PR TITLE
eth/protocols/snap: fix the flaws in the snap sync

### PR DESCRIPTION
**Pull request by rjl493456442**

### CODE REVIEW

File: eth/protocols/snap/sync_test.go [hunk 1]
```diff
@@ -280,48 +286,40 @@ func defaultCodeRequestHandler(t *testPeer, id uint64, hashes []common.Hash, max
...
 			}
 		}
 		hashes = append(hashes, keys)
 		slots = append(slots, vals)
 
-		if incomplete {
+		// Generate the Merkle proofs for the first and last storage slot, but
+		// only if the response was capped. If the entire storage trie included
+		// in the response, no need for any proofs.
+		if originHash != (common.Hash{}) || size >= max {
```
Comment (holiman) [inline eth/protocols/snap/sync_test.go:322]: There exist a cornecase where `size > max` _and_ the response is complete. Hence why I set `incomplete=true` when `limitExceeded=true` _and_ I knew there were more items that could be sent.

**Review by holiman — COMMENTED** [review]
One nitpick

**Comment by holiman** [comment]
One thing I've been testing with (no issues found so far) is adding a `	verifyTrie(syncer.db, sourceAccountTrie.Hash(), t)` in the cases where sync should pass. And add this:

```golang

func verifyTrie(db ethdb.KeyValueStore, root common.Hash, t *testing.T) {
	triedb := trie.NewDatabase(db)
	accTrie, err := trie.New(root, triedb)
	if err != nil {
		t.Fatal(err)
	}
	accounts, slots := 0, 0
	accIt := trie.NewIterator(accTrie.NodeIterator(nil))
	for accIt.Next() {
		var acc struct {
			Nonce    uint64
			Balance  *big.Int
			Root     common.Hash
			CodeHash []byte
		}
		if err := rlp.DecodeBytes(accIt.Value, &acc); err != nil {
			log.Crit("Invalid account encountered during snapshot creation", "err", err)
		}
		accounts++
		if acc.Root != emptyRoot {
			storeTrie, err := trie.NewSecure(acc.Root, triedb)
			if err != nil {
				t.Fatal(err)
			}
			storeIt := trie.NewIterator(storeTrie.NodeIterator(nil))
			for storeIt.Next() {
				slots++
			}
			if err := storeIt.Err; err != nil {
				t.Fatal(err)

			}
		}
	}
	if err := accIt.Err; err != nil {
		t.Fatal(err)
	}
	t.Logf("accounts: %d, slots: %d", accounts, slots)
}
```

File: eth/protocols/snap/sync_test.go (hunk 1 above)
Comment (rjl493456442) [reply eth/protocols/snap/sync_test.go:322]: Yes, it's possible. But I think the downside is OK, we only generate the "unnecessary" proof for this special case in the tests. I think it's acceptable for the simpler code perhaps?

But I'm fine to add it back.

**Review by holiman — APPROVED** [review]

File: eth/protocols/snap/sync_test.go (hunk 1 above)
Comment (rjl493456442) [reply eth/protocols/snap/sync_test.go:320]: The "unnecessary" proof is also handled properly. So I think we can keep it :P

**Comment by holiman** [comment]
Should we fix the sender/part too? 
```diff
diff --git a/eth/protocols/snap/handler.go b/eth/protocols/snap/handler.go
index b9515b8a39..0b4f83a75e 100644
--- a/eth/protocols/snap/handler.go
+++ b/eth/protocols/snap/handler.go
@@ -256,8 +256,13 @@ func handleMessage(backend Backend, peer *Peer) error {
 			var (
 				storage []*StorageData
 				last    common.Hash
+				aborted = false
 			)
-			for it.Next() && size < hardLimit {
+			for it.Next() {
+				if size > hardLimit {
+					aborted = true
+					break
+				}
 				hash, slot := it.Hash(), common.CopyBytes(it.Slot())
 
 				// Track the returned interval for the Merkle proofs
@@ -280,7 +285,7 @@ func handleMessage(backend Backend, peer *Peer) error {
 			// Generate the Merkle proofs for the first and last storage slot, but
 			// only if the response was capped. If the entire storage trie included
 			// in the response, no need for any proofs.
-			if origin != (common.Hash{}) || size >= hardLimit {
+			if origin != (common.Hash{}) || aborted {
 				// Request started at a non-zero hash or was capped prematurely, add
 				// the endpoint Merkle proofs
 				accTrie, err := trie.New(req.Root, backend.Chain().StateCache().TrieDB())

```

**Comment by rjl493456442** [comment]
@holiman I think if the last slot exceeds the limit, we should also generate the proof

```diff
diff --git a/eth/protocols/snap/handler.go b/eth/protocols/snap/handler.go
index b9515b8a3..5478b72cc 100644
--- a/eth/protocols/snap/handler.go
+++ b/eth/protocols/snap/handler.go
@@ -256,8 +256,13 @@ func handleMessage(backend Backend, peer *Peer) error {
 			var (
 				storage []*StorageData
 				last    common.Hash
+				abort   bool
 			)
-			for it.Next() && size < hardLimit {
+			for it.Next() {
+				if size >= hardLimit {
+					abort = true
+					break
+				}
 				hash, slot := it.Hash(), common.CopyBytes(it.Slot())
 
 				// Track the returned interval for the Merkle proofs
@@ -271,6 +276,7 @@ func handleMessage(backend Backend, peer *Peer) error {
 				})
 				// If we've exceeded the request threshold, abort
 				if bytes.Compare(hash[:], limit[:]) >= 0 {
+					abort = true
 					break
 				}
 			}
@@ -280,7 +286,7 @@ func handleMessage(backend Backend, peer *Peer) error {
 			// Generate the Merkle proofs for the first and last storage slot, but
 			// only if the response was capped. If the entire storage trie included
 			// in the response, no need for any proofs.
-			if origin != (common.Hash{}) || size >= hardLimit {
+			if origin != (common.Hash{}) || abort {
 				// Request started at a non-zero hash or was capped prematurely, add
 				// the endpoint Merkle proofs
 				accTrie, err := trie.New(req.Root, backend.Chain().StateCache().TrieDB())

```

File: eth/protocols/snap/sync.go [hunk 2]
```diff
@@ -1553,6 +1553,12 @@ func (s *Syncer) processAccountResponse(res *accountResponse) {
 	// Ensure that the response doesn't overflow into the subsequent task
 	last := res.task.Last.Big()
 	for i, hash := range res.hashes {
+		// Mark the range complete if the last is already included.
+		// Keep iteration to delete the extra states if exists.
+		if hash.Big().Cmp(last) == 0 {
```
Comment (karalabe) [inline eth/protocols/snap/sync.go:1558]: Nitpick, `hash.Big().Cmp(last)` will almost never be 0, so we'll almost always redo the exact same calculation below. Might as well be a bit smarter?

```go
cmp := hash.Big().Cmp(last) 
if cmp == 0 {
...
}
if cmp > 0 {
...
}
```

?

File: eth/protocols/snap/sync.go [hunk 3]
```diff
@@ -1761,6 +1767,12 @@ func (s *Syncer) processStorageResponse(res *storageResponse) {
 				// Ensure the response doesn't overflow into the subsequent task
 				last := res.subTask.Last.Big()
 				for k, hash := range res.hashes[i] {
+					// Mark the range complete if the last is already included.
+					// Keep iteration to delete the extra states if exists.
+					if hash.Big().Cmp(last) == 0 {
+						res.cont = false
+						continue
+					}
```
Comment (karalabe) [inline eth/protocols/snap/sync.go:1775]: Same here wrt the double big+cmp calls

File: eth/protocols/snap/sync_test.go (hunk 1 above)
Comment (rjl493456442) [reply eth/protocols/snap/sync_test.go]: Fixed

**Review by holiman — APPROVED** [review]
J'approuf

**Review by karalabe — APPROVED** [review]
LGTM

**Review by alisscco — COMMENTED** [review]
👌🏻

**Pull request merged**